src/java.base/share/classes/java/security/PEMEncoder.java

```diff
@@ -0,0 +1,358 @@
+     */
+    private PEMEncoder(PBEKeySpec pbe) {
+        keySpec = pbe;
+        lock = new ReentrantLock();
+    }
+
+    /**
+     * Returns an instance of PEMEncoder.
+     *
+     * @return PEMEncoder instance
+     */
+    static public PEMEncoder of() {
+        return PEM_ENCODER;
+    }
+
+    /**
+     * Construct a String-based encoding based off the id type.
+     * @return the string
+     */
+    private String pemEncoded(PEMRecord pem) {
+        StringBuilder sb = new StringBuilder(1024);
+        sb.append("-----BEGIN ").append(pem.id()).append("-----");
+        sb.append(System.lineSeparator());
+        if (b64Encoder == null) {
+            b64Encoder = Base64.getMimeEncoder(64,
+                System.lineSeparator().getBytes());
+        }
+        sb.append(b64Encoder.encodeToString(
+            pem.pem().getBytes(StandardCharsets.ISO_8859_1)));
+        sb.append(System.lineSeparator());
+        sb.append("-----END ").append(pem.id()).append("-----");
+        sb.append(System.lineSeparator());
+        return sb.toString();
+    }
+
+    /**
+     * Encoded a given {@code DEREncodable} and return the PEM encoding in a
+     * String
+     *
+     * @param de a cryptographic object to be PEM encoded that implements
+     *           DEREncodable.
+     * @return PEM encoding in a String
+     * @throws IllegalArgumentException when the passed object returns a null
+     * binary encoding. An exception is thrown when PEMEncoder is
+     * configured for encryption while encoding a DEREncodable that does
+     * not support encryption.
+     * @throws NullPointerException when object passed is null.
+     * @see #withEncryption(char[])
+     */
+    public String encodeToString(DEREncodable de) {
+        Objects.requireNonNull(de);
+        return switch (de) {
+            case PublicKey pu -> buildKey(null, pu.getEncoded());
+            case PrivateKey pr -> buildKey(pr.getEncoded(), null);
+            case KeyPair kp -> {
+                if (kp.getPublic() == null) {
+                    throw new IllegalArgumentException("KeyPair does not " +
+                        "contain PublicKey.");
+                }
+
+                if (kp.getPrivate() == null) {
+                    throw new IllegalArgumentException("KeyPair does not " +
+                        "contain PrivateKey.");
+                }
+                yield buildKey(kp.getPrivate().getEncoded(),
+                    kp.getPublic().getEncoded());
+            }
+            case X509EncodedKeySpec x ->
+                buildKey(null, x.getEncoded());
+            case PKCS8EncodedKeySpec p ->
+                buildKey(p.getEncoded(), null);
+            case EncryptedPrivateKeyInfo epki -> {
+                try {
+                    yield pemEncoded(new PEMRecord(
+                        PEMRecord.ENCRYPTED_PRIVATE_KEY, epki.getEncoded()));
+                } catch (IOException e) {
+                    throw new SecurityException(e);
```

Comment:
Do you really want to use `SecurityException`? This only happens when the `AlgorithmParameters` inside EPKI is not initialized. I would say this is a very good candidate for an `IllegalStateException`.